django/db/migrations/autodetector.py

```diff
@@ -1352,10 +1353,20 @@ def create_altered_constraints(self):
             old_model_state = self.from_state.models[app_label, old_model_name]
             new_model_state = self.to_state.models[app_label, model_name]
 
-            old_constraints = old_model_state.options[option_name]
-            new_constraints = new_model_state.options[option_name]
-            add_constraints = [c for c in new_constraints if c not in old_constraints]
-            rem_constraints = [c for c in old_constraints if c not in new_constraints]
+            old_constraints = {
+                serializer_factory(c).serialize()[0]: c
+                for c in old_model_state.options[option_name]
+            }
+            new_constraints = {
+                serializer_factory(c).serialize()[0]: c
+                for c in new_model_state.options[option_name]
+            }
+            add_constraints = [
+                c for k, c in new_constraints.items() if k not in old_constraints
+            ]
+            rem_constraints = [
+                c for k, c in old_constraints.items() if k not in new_constraints
+            ]
```

Comment:
What do you think about improving `__eq__` and `__hash__` for `Q` objects instead? `Expression` has a custom implementation:
https://github.com/django/django/blob/0f3b1a783dfa36cb23aae0bb954756d0edcd9fc1/django/db/models/expressions.py#L478-L503
Something similar for `Q` objects should fix this issue, e.g.:
```diff
diff --git a/django/db/models/query_utils.py b/django/db/models/query_utils.py
index 78148f76b0..c0f1d4faa2 100644
--- a/django/db/models/query_utils.py
+++ b/django/db/models/query_utils.py
@@ -14,6 +14,8 @@ from django.core.exceptions import FieldError
 from django.db import DEFAULT_DB_ALIAS, DatabaseError, connections
 from django.db.models.constants import LOOKUP_SEP
 from django.utils import tree
+from django.utils.functional import cached_property
+from django.utils.hashable import make_hashable
 
 logger = logging.getLogger("django.db.models")
 
@@ -151,6 +153,23 @@ class Q(tree.Node):
             kwargs["_negated"] = True
         return path, args, kwargs
 
+    @cached_property
+    def identity(self):
+        path, args, kwargs = self.deconstruct()
+        identity = [path, *kwargs.items()]
+        for arg, value in args:
+            value = make_hashable(value)
+            identity.append((arg, value))
+        return tuple(identity)
+
+    def __eq__(self, other):
+        if not isinstance(other, Q):
+            return NotImplemented
+        return other.identity == self.identity
+
+    def __hash__(self):
+        return hash(self.identity)
+
 
 class DeferredAttribute:
     """
```